Sources/OpenCombine/Publishers/Publishers.FlatMap.swift

```diff
@@ -0,0 +1,402 @@
+        private var sendFinishedAfterDrainingQueue = false
+        private var upstreamSubscription: Subscription?
+
+        var description: String { return "FlatMap" }
+
+        init(downstream: Downstream,
+             maxPublishers: Subscribers.Demand,
+             transform: @escaping (Upstream.Output) -> Child) {
+            self.downstream = downstream
+            self.maxPublishers = maxPublishers
+            self.transform = transform
+        }
+
+        final func cancel() {
+            let (upstreamToCancel, childrenToCancel) = lock.do { ()
+                -> (Subscription?, Set<ChildSubscriber>) in
+                let upstreamToCancel = upstreamSubscription
+                upstreamSubscription = nil
+                return (upstreamToCancel, lockedDeactivateAndReturnChildToCancel())
+            }
+            upstreamToCancel?.cancel()
+            cancelChildren(childrenToCancel)
+        }
+    }
+}
+
+// Private implementation
+extension Publishers.FlatMap.Inner {
+    private func deactivate() {
+        cancelChildren(lock.do(lockedDeactivateAndReturnChildToCancel))
+    }
+
+    // Must be called with lock held.
+    private func lockedDeactivateAndReturnChildToCancel() -> Set<ChildSubscriber> {
+        downstream = nil
+        downstreamDemand = .none
+        let result = childSubscribers
+        childSubscribers.removeAll()
+        upstreamSubscription = nil
+        return result
+    }
+
+    private func cancelChildren(_ childrenToCancel: Set<ChildSubscriber>) {
+        childrenToCancel.forEach { $0.cancel() }
+    }
+
+    private final func evaluateActiveSubscriptionsExecuting(lockedWork: () -> Void) {
+        let shouldProcessQueue: Bool = lock.do {
+            lockedWork()
+            if childSubscribers.isEmpty && upstreamSubscription == nil {
+                sendFinishedAfterDrainingQueue = true
+                if !queueIsBeingProcessed {
+                    queueIsBeingProcessed = true
+                    return true
+                }
+            }
+            return false
+        }
+
+        if shouldProcessQueue {
+            processQueue()
+        }
+    }
+
+    private func child(_ child: ChildSubscriber,
+                       completedWith completion: Subscribers.Completion<Failure>) {
+        switch completion {
+        case .finished:
+            removeActiveSubscription(forChild: child)
+        case .failure:
+            downstream?.receive(completion: completion)
+            deactivate()
+        }
+    }
+
+    private func removeActiveSubscription(forChild child: ChildSubscriber) {
+        evaluateActiveSubscriptionsExecuting { childSubscribers.remove(child) }
+    }
+
+    private func child(_ child: ChildSubscriber,
+                       receivedValue value: Child.Output) -> Subscribers.Demand {
+        // Obtain the lock in order to determine what to do
+        let (valueWasBuffered, processTheQueue): (Bool, Bool) = lock.do {
```

Comment:
I'm having trouble understanding the logic in this block, can you add a comment?

Reply:
Yeah. This is the same code from above. I'll add more documentation after I fix the potential bug you uncovered.